Update Rodio to 0.21

mnmaita opened the pull request:
# Objective

- Closes #19672 

## Solution

- Updated both `cpal` and `rodio` to their latest versions.
- Updated code to address `rodio`'s breaking changes.
- Reworked audio related feature flags so `symphonia` support becomes the default.

## Testing

- Tested audio related examples.
- NOTE: `soundtrack` and spatial examples are not working in this PR yet.

mnmaita commented on crates/bevy_audio/src/audio_source.rs:
```diff
@@ -80,22 +81,16 @@ impl AssetLoader for AudioLoader {
+    /// which iterates over samples of type [`rodio::Sample`].
     /// Must be a [`rodio::Source`] so that it can provide information on the audio it is iterating over.
-    type Decoder: rodio::Source + Send + Iterator<Item = Self::DecoderItem>;
+    type Decoder: rodio::Source + Send + Iterator<Item = rodio::Sample>;
```
`rodio::Source` has `Iterator<Item = rodio::Sample>` as a bound as far as I can see, should we drop this redundant bound or should we keep it?

mnmaita commented:
There's an unreleased change in `cpal` that makes `Source` `Send`. This will fix the build issues in MacOS as far as I can tell. I'll try bothering some people here and there to push this forward... 😅

mnmaita commented:
> We shouldn't be exposing `minimp3` at all, per #20183

Rebased, removed the unwanted feature and fixed some wrongly set internal feature flags for the fallbacks. I've also cleaned up my own messages as there was too much noise in the PR. Later I will try pushing a temporary commit to see if I can fix the outstanding issues and we can finally test this, analyze what options we have and make a decision for this upgrade.

mnmaita commented:
> Can you track down the `cpal` authors and pester them (politely and patiently!) for a release? You're correct that we can't ship this with a git dependency.

Relevant: https://github.com/RustAudio/cpal/pull/1065 😄 

Please also check this question I had:

> Is it ok to have lewton as the default ogg backend due to https://github.com/RustAudio/rodio/issues/775#issuecomment-3157459818 with the symphonia implementation when using looped/buffered sources?

It seems like a PR that fixed that "got too big" and was just closed, and I didn't see any other PRs pushed to fix that. So as a follow up question:

> Do we want to make symphonia the default (for the rest of the audio formats)?

As far as I understand, symphonia is an opt-in feature right now.

Thanks @alice-i-cecile!

mnmaita commented on examples/mobile/build_rust_deps.sh:
```diff
@@ -54,7 +54,7 @@ for arch in $ARCHS; do
       else
         # M1 iOS simulator
-        TARGET=aarch64-apple-ios-sim
+        TARGET=arm64-apple-ios-simulator
```
You are completely right, I've reverted the change. I've checked dependencies (`cpal` and `coreaudio-rs` specifically) and they don't depend on `bindgen` any longer as far as I could see, so the issue should be solved but we can always reopen https://github.com/bevyengine/bevy/issues/18893 if any unexpected issues arise. Thanks for the catch! Let me know if this looks ok now.

rparrett commented on _release-content/migration-guides/rodio_0_22.md:
```diff
@@ -0,0 +1,44 @@
+- `symphonia-vorbis`
+- `symphonia-wav`
+
+Notice that OGG/VORBIS support through `symphonia` is currently subject to issues with buffering, reverb, looping and spatial audio. Check the following issues/PRs for additional context:
```
ditto

rparrett commented on docs-template/EXAMPLE_README.md.tpl:
````diff
@@ -160,10 +160,12 @@ In its examples, Bevy targets the minimum Android API that Play Store  <!-- mark
 
 ```toml
-bevy = { version = "0.14", default-features = false, features = ["android-native-activity", ...] }
+bevy = { version = "0.16", default-features = false, features = ["android-native-activity", ...] }
````
```suggestion
bevy = { version = "0.19", default-features = false, features = ["android-native-activity", ...] }
```
😅

rparrett commented:
Pretty sure CI needs to be updated to run cargo ndk with `-P 26` or something along those lines. That's the same error I was seeing locally.

But I'm assuming this might cause issues in other automation (the example runner, etc)...

mnmaita commented:
> > I've added said minimum API version note in the Android docs and the release notes. It should only apply when using `bevy_audio` or the default bevy features, right? Please let me know if I should reword that or add that somewhere else.
> 
> Looks good, thanks!
> 
> > More importantly, should I make the change suggested by @rparrett then? Not sure if we have to make other changes elsewhere though.
> 
> Yup! And there too: https://github.com/bevyengine/bevy-example-runner/blob/main/.github/workflows/workflow-mobile.yml#L148

Rebased, done and... done! Thanks for taking a look at this. 🙏🏻 

Btw should we also update [this one](https://github.com/bevyengine/bevy/blob/main/docs-template/EXAMPLE_README.md.tpl#L110) and [this other one](https://github.com/bevyengine/bevy/blob/main/docs-template/EXAMPLE_README.md.tpl#L116) in that readme template?
